src/compiler/checker.ts

```diff
@@ -22628,11 +22628,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             let i = 0;
             for (const type of target.types) {
                 const targetType = getTypeOfPropertyOfType(type, propertyName);
-                if (targetType && related(getDiscriminatingType(), targetType)) {
-                    discriminable[i] = discriminable[i] === undefined ? true : discriminable[i];
-                }
-                else {
-                    discriminable[i] = false;
+                if (discriminable[i] === undefined) {
+                    discriminable[i] = !!(targetType && related(getDiscriminatingType(), targetType));

```

Comment:
What happens if you switch this to check `discriminable[i] !== true` so that as long as *some* type has a successful discriminant, the property is discriminable?

Or alternatively,

```ts
checkEachType: for (const type of target.types) {
    const targetType = getTypeOfPropertyOfType(type, propertyName);

    const targetIsDiscriminable = discriminable[i] = !!(targetType && related(getDiscriminatingType(), targetType));
    if (targetIsDiscriminable) {
        break checkEachType;
    }

    i++
}
```

Reply:
That doesn't change the logic though, it just optimizes the looping a bit.

Reply:
I do like your suggestion BTW, just making sure I'm not missing something.

Reply:
Doesn't it?

In the original version, a single `false` takes precedence over everything else.

With your change, the first result takes precedence over every other (which sounds suspicious - this doesn't even need to be a loop anymore).

In my suggestion, a single `true` takes precedence over any other result.

Reply:
Right, that makes sense. Boolean logic gets tricky with truthy and falsy stuff.

Reply:
Also, with my change the first result doesn't always take precedence, because the objects in the union could potentially contain different items and so the writes to `discriminable` could differ between iterations.

Reply:
@DanielRosenwasser your suggested change breaks [this test](https://github.com/microsoft/TypeScript/blob/e83d61398ea0e4231e882121dd6c6bcfe4fdc9e4/tests/cases/compiler/contextualTypeShouldBeLiteral.ts#L40). 
```
foo2({
  type2: 'y',
  ~~~~~
!!! error TS2345: Argument of type '{ type2: string; value: string; method(): void; }' is not assignable to parameter of type 'X2 | Y2'.
!!! error TS2345:   Object literal may only specify known properties, but 'type2' does not exist in type 'X2'. Did you mean to write 'type1'?
  value: 'done',
  method() {
    this;
    this.value;
  }
});
```